osgar/test_zmqrouter.py

```diff
@@ -0,0 +1,252 @@
+            assert value == i
+            time.sleep(0.15)
+            dt = self.bus.publish(self.output_name, i)
+            #print(self.bus.name, dt, channel, value)
+
+class NoQuit:
+    def __init__(self, config, bus):
+        self.bus = bus
+        self.bus.register()
+
+    def start(self):
+        self.thread = Thread(target=self.run)
+        self.thread.start()
+
+    def join(self, timeout=None):
+        self.thread.join(timeout)
+
+    def run(self):
+        while True:
+            time.sleep(1)
+
+
+class Test(unittest.TestCase):
+    def setUp(self):
+        self.tempdir = pathlib.Path(tempfile.mkdtemp(dir=pathlib.Path.cwd()))
+        os.environ['OSGAR_LOGS'] = str(self.tempdir)
+
+    def tearDown(self):
+        shutil.rmtree(self.tempdir)
+
+    def test_threads(self):
+        main()
+
+    def test_noop(self):
+        config = {
+            'version': 2,
+            'robot': {
+                'modules': {
+                    "noop": {
+                        "driver": "osgar.test_zmqrouter:Noop",
+                        "init": {}
+                    },
+                }, 'links':[]
+            }
+        }
+        record(config, log_filename='noop.log')
+
+    def test_publisher_single(self):
+        config = { 'version': 2, 'robot': { 'modules': {}, 'links': [] } }
+        config['robot']['modules']['publisher'] = {
+            "driver": "osgar.test_zmqrouter:Publisher",
+            "init": { "output": "count"}
+        }
+        record(config, log_filename='publisher.log')
+        with osgar.logger.LogReader(self.tempdir/"publisher.log", only_stream_id=1) as log:
+            last_dt = datetime.timedelta()
+            count = 0
+            for dt, channel, data in log:
+                self.assertGreater(dt, last_dt)
+                self.assertEqual(int.from_bytes(data, 'little'), count)
+                last_dt = dt
+                count += 1
+
+    def test_publisher_multi(self):
+        config = { 'version': 2, 'robot': { 'modules': {}, 'links': [] } }
+        for i in range(3):
+            config['robot']['modules'][f'publisher{i}'] = {
+                "driver": "osgar.test_zmqrouter:Publisher",
+                "init": { "output": f"count{i}" }
+            }
+        record(config, log_filename='publisher.log')
+
+    def test_compress(self):
+        config = { 'version': 2, 'robot': { 'modules': {}, 'links': [] } }
+        config['robot']['modules']['publisher'] = {
+            "driver": "osgar.test_zmqrouter:Publisher",
+            "init": { "output": "count:gz"}
+        }
+        record(config, log_filename='compressed-publisher.log')
+
+    def test_null(self):
+        config = { 'version': 2, 'robot': { 'modules': {}, 'links': [] } }
+        config['robot']['modules']['publisher'] = {
+            "driver": "osgar.test_zmqrouter:Publisher",
+            "init": { "output": "count:null"}
+        }
+        record(config, log_filename='null-publisher.log')
```

Comment:
the `record` used to be by default infinite, which is probably not the case here?

Reply:
The whole recording session ends when one of the nodes calls `request_stop`, addressing issue #245. Each of the nodes call `request_top` automatically from `child`.